lib/githubactions/token_validator.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2022 Gravitational, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package githubactions
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/coreos/go-oidc"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+type IDTokenValidatorConfig struct {
+	// Clock is used by the validator when checking expiry and issuer times of
+	// tokens. If omitted, a real clock will be used.
+	Clock clockwork.Clock
+	// IssuerURL is the URL of the OIDC token issuer, on which the
+	// /well-known/openid-configuration endpoint can be found.
+	// If this is omitted, a default value will be set.
+	IssuerURL string
+}
+
+type IDTokenValidator struct {
+	IDTokenValidatorConfig
+
+	mu    sync.Mutex
+	_oidc *oidc.Provider
```

Comment:
I really don't like the `_` here. Is this a new convention that I might be unaware of?

Reply:
Hm, no, I was more trying to indicate it was a field protected by a mutex, but I'll use a comment instead.

Reply:
773926ea1